lib/fastifySession.js

```diff
@@ -106,54 +110,49 @@ function session (fastify, options, next) {
         done()
         return
       }
-      let sessionId = request.cookies[cookieName]
-      if (sessionId && hasCookiePrefix && sessionId.startsWith(cookiePrefix)) {
-        sessionId = sessionId.slice(cookiePrefixLength)
-      }
-      if (!sessionId) {
+
+      const cookieSessionId = getCookieSessionId(request)
+      if (!cookieSessionId) {
         request.session = new Session(request, idGenerator, cookieOpts, cookieSigner)
         done()
       } else {
-        decryptSession(sessionId, options, request, done)
+        decryptSession(cookieSessionId, options, request, done)
       }
     }
   }
 
   function onSend (options) {
     const cookieOpts = options.cookie
-    const cookieName = options.cookieName
-    const cookiePrefix = options.cookiePrefix
-    const saveUninitialized = options.saveUninitialized
-    const hasCookiePrefix = typeof cookiePrefix === 'string' && cookiePrefix.length !== 0
+    const saveUninitializedSession = options.saveUninitialized
+    const rollingSessions = options.rolling
 
     return function saveSession (request, reply, payload, done) {
       const session = request.session
-      if (!session || !session.sessionId) {
+      if (!session || !session.sessionId || !session.encryptedSessionId) {
         done()
         return
       }
 
-      let encryptedSessionId = session.encryptedSessionId
-      if (encryptedSessionId && hasCookiePrefix) {
-        encryptedSessionId = `${cookiePrefix}${encryptedSessionId}`
-      }
-
-      if (!shouldSaveSession(request, cookieOpts, saveUninitialized)) {
-      // if a session cookie is set, but has a different ID, clear it
-        if (request.cookies[cookieName] && request.cookies[cookieName] !== encryptedSessionId) {
+      const cookieSessionId = getCookieSessionId(request)
+      const shouldSaveSession = getShouldSaveSession(request, cookieSessionId, saveUninitializedSession, rollingSessions)
+      const isInsecureConnection = cookieOpts.secure === true && !isConnectionSecure(request)
```

Comment:
cookieOpts.secure can be also have the value 'auto'. I dont know if this is a good idea to just check for secure being true. Also secure could be set to undefined. So I have some doubts if this code is working as expected under any circumstances.

Reply:
Yeah, I think this is a potential issue that already exists in the code. Probably for another PR. Looking at:

![image](https://user-images.githubusercontent.com/736775/187939670-3dc5c2a4-bf88-4c62-a9cb-bf8acb0e71c0.png)

If you add a `!`:

```
!(cookieOpts.secure !== true || cookieOpts.secure === 'auto')
(cookieOpts.secure === true && cookieOpts.secure !== 'auto')
(cookieOpts.secure === true)
```

I think we should fix this in a separate ticket.

Reply:
We should create an issue after we merged this